File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/DataTypePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;

    class DataTypePersistence
    {
        public static CdmDataTypeDefinition FromData(CdmCorpusContext ctx, JToken obj)
        {
            if (obj == null)
            {
                return null;
            }
            CdmDataTypeDefinition dataType = ctx.Corpus.MakeObject<CdmDataTypeDefinition>(CdmObjectType.DataTypeDef, (string)obj["dataTypeName"]);
            dataType.ExtendsDataType = DataTypeReferencePersistence.FromData(ctx, obj["extendsDataType"]);

            if (obj["explanation"] != null)
                dataType.Explanation = (string)obj["explanation"];

            Utils.AddListToCdmCollection(dataType.ExhibitsTraits, Utils.CreateTraitReferenceList(ctx, obj["exhibitsTraits"]));
            return dataType;
        }

        public static DataType ToData(CdmDataTypeDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            return new DataType
            {
                Explanation = instance.Explanation,
                DataTypeName = instance.DataTypeName,
                ExtendsDataType =  Utils.JsonForm(instance.ExtendsDataType, resOpt, options),
                ExhibitsTraits = CopyDataUtils.ListCopyData(resOpt, instance.ExhibitsTraits, options)
            };
        }
    }
}

File: objectModel/Python/cdm/persistence/syms/data_type_persistence.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from cdm.objectmodel import CdmCorpusContext, CdmDataTypeDefinition
from cdm.enums import CdmObjectType
from cdm.utilities import ResolveOptions, CopyOptions, copy_data_utils

from . import utils
from .types import DataType
from .data_type_reference_persistence import DataTypeReferencePersistence


class DataTypePersistence:
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: DataType) -> CdmDataTypeDefinition:
        if data is None:
            return None

        data_type = ctx.corpus.make_object(CdmObjectType.DATA_TYPE_DEF, data.dataTypeName)
        data_type.extends_data_type = DataTypeReferencePersistence.from_data(ctx, data.get('extendsDataType'))
        data_type.explanation = data.get('explanation')

        utils.add_list_to_cdm_collection(data_type.exhibits_traits,
                                         utils.create_trait_reference_array(ctx, data.get('exhibitsTraits')))

        return data_type

    @staticmethod
    def to_data(instance: CdmDataTypeDefinition, res_opt: ResolveOptions, options: CopyOptions) -> DataType:
        result = DataType()
        result.explanation = instance.explanation
        result.dataTypeName = instance.data_type_name
        result.extendsDataType = DataTypeReferencePersistence.to_data(instance.extends_data_type, res_opt, options) if instance.extends_data_type else None
        result.exhibitsTraits = copy_data_utils._array_copy_data(res_opt, instance.exhibits_traits, options)
        return result
